### PR TITLE
[SOL-1356] Create Coupon UI

### PR DESCRIPTION
**Pull request by dsego**
Implements a UI for creating new coupons that models after the course admin UI.

@mjfrey, @clintonb

### CODE REVIEW

File: ecommerce/coupons/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    path = reverse('coupons:app', args=[''])
+
+    def test_login_required(self):
+        """ Users are required to login before accessing the view. """
+        self.client.logout()
+        response = self.client.get(self.path)
+        self.assertEqual(response.status_code, 302)
+        self.assertIn(settings.LOGIN_URL, response.url)
+
+    @httpretty.activate
```
Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py]: This is not needed.

File: ecommerce/coupons/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    def test_staff_user_required(self):
+        """ Verify the view is only accessible to staff users. """
+        user = self.create_user(is_staff=False)
+        self.client.login(username=user.username, password=self.password)
+        response = self.client.get(self.path)
+        self.assertEqual(response.status_code, 404)
+
+        user = self.create_user(is_staff=True)
+        self.create_access_token(user)
```
Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py]: Why do you need an access token?

File: ecommerce/coupons/views.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from django.contrib.auth.decorators import login_required
+from django.utils.decorators import method_decorator
+from django.views.generic import TemplateView
+from django.http import Http404
+
+
+class StaffOnlyMixin(object):
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: This already exists in courses/views. Move it to core/views and share it.

File: ecommerce/static/js/apps/coupon_admin_app.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        /**
+         * Navigate to a new page within the app.
+         *
+         * Attempts to open the link in a new tab/window behave as the user expects, however the app
+         * and data will be reloaded in the new tab/window.
+         *
+         * @param {Event} event - Event being handled.
+         * @returns {boolean} - Indicates if event handling succeeded (always true).
+         */
+        navigate = function (event) {
```
Comment (clintonb) [inline ecommerce/static/js/apps/coupon_admin_app.js]: Don't duplicate code. Find a way to share it.

File: ecommerce/static/js/collections/coupon_collection.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    function (Backbone,
+              _,
+              CouponModel) {
+        'use strict';
+
+        return Backbone.Collection.extend({
+            model: CouponModel,
+            url: '/api/v2/coupons/',
+
+            parse: function (response) {
```
Comment (clintonb) [inline ecommerce/static/js/collections/coupon_collection.js]: Again, stop duplicating code. Create a new class, `PaginatedCollection`, if necessary.

File: ecommerce/static/js/models/coupon_model.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
+// jscs:disable requireCapitalizedConstructors
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Why is this disabled? If you really need to disable a validation, do so only for the line(s) where necessary.

File: ecommerce/static/js/models/coupon_model.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+              $,
+              $cookie,
+              _,
+              moment,
+              Course) {
+        'use strict';
+
+        _.extend(Backbone.Model.prototype, Backbone.Validation.mixin);
+
+        _.extend(Backbone.Validation.patterns, {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Move to a mixin shared with the Course model.

File: ecommerce/static/js/models/coupon_model.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        });
+
+        _.extend(Backbone.Validation.messages, {
+            courseId: gettext('The course ID is invalid.')
+        });
+
+        return Backbone.RelationalModel.extend({
+            urlRoot: '/api/v2/coupons/',
+
+            defaults: {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js:26]: Remove.

File: ecommerce/static/js/models/coupon_model.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+            isEnrollmentCode: function () {
+                return this.get('code_type') === 'enrollment' ;
+            },
+
+            isDiscountCode: function () {
+                return this.get('code_type') === 'discount' ;
+            },
+
+            labels: {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Remove.

File: ecommerce/static/js/models/coupon_model.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                },
+            ],
+
+            initialize: function () {
+                this.on('change:voucher_type', this.changeVoucherType, this);
+                this.on('change:course_id', this.loadCourse, this);
+                this.set('quantity', 1);
+            },
+
+            loadCourse: function (model, value) {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Add a docstring.

File: ecommerce/static/js/models/coupon_model.js [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                if (this.isValid('course_id')) {
+                    this.set('course', Course.findOrCreate({id: value}));
+                    this.listenTo(this.get('course'), 'change', this.fillFromCourse, this);
+                    this.get('course').fetch({data: {include_products: true}});
+                }
+            },
+
+            fillFromCourse: function (course) {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Add a docstring.

File: ecommerce/static/js/models/coupon_model.js [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                        label: val.getSeatTypeDisplayName()
+                    };
+                });
+                this.set('seatTypes', seatTypes);
+
+                // triggers update for seat_type selectOptions
+                this.set('seat_type', seatTypes[0].value);
+            },
+
+            changeVoucherType: function (model, value) {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js:118]: Add a docstring.

File: ecommerce/static/js/models/coupon_model.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+                if (value === 'Single use') {
+                    this.set('quantity', 1);
+                }
+            },
+
+            save: function (options) {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js:124]: Add a docstring.

File: ecommerce/static/js/routers/coupon_router.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        'pages/coupon_create_page'
+    ],
+    function (Backbone,
+              BackboneRouteFilter,
+              BackboneSuper,
+              CouponListPage,
+              CouponCreatePage) {
+        'use strict';
+
+        return Backbone.Router.extend({
```
Comment (clintonb) [inline ecommerce/static/js/routers/coupon_router.js]: Move the duplicate code to a shared parent class.

File: ecommerce/static/js/test/specs/views/coupon_create_view_spec.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+                // quantity
+                view.$el.find('[name=voucher_type]').val('Single use').trigger('change');
+                expect(visible('[name=quantity]')).toBe(true);
+                view.$el.find('[name=voucher_type]').val('Multi-use').trigger('change');
+                expect(visible('[name=quantity]')).toBe(false);
+                view.$el.find('[name=voucher_type]').val('Once per customer').trigger('change');
+                expect(visible('[name=quantity]')).toBe(false);
+
+                function visible(selector) {
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/views/coupon_create_view_spec.js]: Functions should be declared before they are used.

**Comment by clintonb** [comment]
Please include screenshots.

**Comment by dsego** [comment]
@clintonb Awesome feedback, thanks! 
I tried to address everything, please take another look. I will try to improve coverage and add screenshots tomorrow.

**Comment by clintonb** [comment]
@AlasdairSwan @bderusha would you add extra eyes on this, please?

File: ecommerce/static/js/apps/coupon_admin_app.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+        'utils/navigate'
+    ],
+    function (Backbone,
+              CouponRouter,
+              navigate) {
+        'use strict';
+
+        $(function () {
+            var $app = $('#app');
+            var couponApp = new CouponRouter({$el: $app});
```
Comment (AlasdairSwan) [inline ecommerce/static/js/apps/coupon_admin_app.js]: Nitpick: only need 1 var declaration, can comma separate the variables. i.e.

```
var $app = $('#app'),
    couponApp = ....
```

File: ecommerce/static/js/apps/course_admin_app.js [hunk 17]
```diff
@@ -2,66 +2,23 @@ require([
...
-         *
-         * @param {String} fragment
-         */
-        Backbone.View.prototype.goTo = function (fragment) {
-            courseApp.navigate(fragment, {trigger: true});
-        };
-
         $(function () {
             var $app = $('#app');
+            var courseApp = new CourseRouter({$el: $app});
```
Comment (AlasdairSwan) [inline ecommerce/static/js/apps/course_admin_app.js]: Nitpick: only need 1 var declaration, can comma separate the variables.

File: ecommerce/static/js/models/coupon_model.js [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+                },
+                start_date: function (val) {
+                    if (_.isEmpty(val)) {
+                        return gettext('Start date is required');
+                    }
+                    var startDate = moment(new Date(val));
+                    if (!startDate.isValid()) {
+                        return gettext('Start date is invalid');
+                    }
+                    var endDate = moment(this.get('end_date'));
```
Comment (AlasdairSwan) [inline ecommerce/static/js/models/coupon_model.js]: vars should be declared at the top of a function due to hoisting (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var#var_hoisting)

File: ecommerce/static/js/models/coupon_model.js [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+                var data = {
+                    title: this.get('title'),
+                    client_username: this.get('client_username'),
+                    start_date: moment.utc(this.get('start_date')),
+                    end_date: moment.utc(this.get('end_date')),
+                    stock_record_ids: stockRecordIds,
+                    code: this.get('code') || '',
+                    voucher_type: this.get('voucher_type'),
+                    quantity: parseInt(this.get('quantity'))
+                };
```
Comment (AlasdairSwan) [inline ecommerce/static/js/models/coupon_model.js]: vars should be declared at the top of a function due to hoisting (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var#var_hoisting)

**Comment by dsego** [comment]
@clintonb @AlasdairSwan I have done some refactoring to share more code between views. Hope I didn't overcomplicate it.

**Comment by dsego** [comment]
screen shot 2015-12-14 at 10 07 10 am

File: ecommerce/static/js/routers/page_router.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+                return this;
+            },
+
+            /**
+             * 404 page
+             * @param {String} path - Invalid path.
+             */
+            notFound: function (path) {
+                // TODO Render something!
```
Comment (bderusha) [inline ecommerce/static/js/routers/page_router.js:64]: I know this is still in progress, but I wanted to put a comment here so we don't forget about this `TODO`

File: ecommerce/static/js/test/specs/views/form_view_spec.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+                expect(model.save).toHaveBeenCalled();
+                expect(view.$el.find('.alert').length).toBe(1);
+                expect(view.$el.find('.alert').html()).toBe(errorHTML);
+
+                // return responseJSON
+                errorObj = { responseJSON: { error: 'An error occurred while saving the data.' }};
+                view.submit($.Event('click'));
+                expect(model.save).toHaveBeenCalled();
+                expect(view.$el.find('.alert').length).toBe(1);
+                expect(view.$el.find('.alert').html()).toBe(errorHTML);
```
Comment (bderusha) [inline ecommerce/static/js/test/specs/views/form_view_spec.js]: Seems like a lot of this error checking is shared.  Can we pull this out into its own method?

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @dsego! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=460

File: ecommerce/courses/views.py [hunk 22]
```diff
@@ -16,19 +14,11 @@
 
 from ecommerce.extensions.partner.shortcuts import get_partner_for_site
 from ecommerce.settings import get_lms_url
+from ecommerce.core.views import StaffOnlyMixin
```
Comment (clintonb) [inline ecommerce/courses/views.py]: Alphabetize.

File: ecommerce/static/js/collections/coupon_collection.js [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+    function (Backbone,
+              BackboneSuper,
+              _,
+              PaginatedCollection,
+              CouponModel) {
+        'use strict';
+
+        return PaginatedCollection.extend({
+            model: CouponModel,
+            url: '/api/v2/products/',
```
Comment (clintonb) [inline ecommerce/static/js/collections/coupon_collection.js:17]: We have a couple thousand products in production. Make an endpoint explicitly for coupons, or add some form of filtering to the products endpoint.

File: ecommerce/static/js/models/coupon_model.js [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+
+            isEnrollmentCode: function () {
+                return this.get('code_type') === 'enrollment' ;
+            },
+
+            isDiscountCode: function () {
+                return this.get('code_type') === 'discount' ;
+            },
+
+            relations: [
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Why is this necessary? This will complicate the move to dynamic catalogs.

File: ecommerce/static/js/models/coupon_model.js [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+                    key: 'course',
+                    relatedModel: Course,
+                    includeInJSON: false
+                },
+            ],
+
+            initialize: function () {
+                this.on('change:voucher_type', this.changeVoucherType, this);
+                this.on('change:course_id', this.loadCourse, this);
+                this.set('quantity', 1);
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Use [defaults](http://backbonejs.org/#Model-defaults).

File: ecommerce/static/js/models/coupon_model.js [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+                    this.set('course', Course.findOrCreate({id: value}));
+                    this.listenTo(this.get('course'), 'change', this.fillFromCourse, this);
+                    this.get('course').fetch({data: {include_products: true}});
+                }
+            },
+
+            /**
+             * Fill seat type options from course data.
+             */
+            fillFromCourse: function (course) {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: This strikes me as a code smell. On the backend, coupons aren't directly related to courses. It seems you are only making the distinction to update the view. If that is the case, courses should not be intermingled with your coupon model. Regardless, you should not be creating labels in a model.

File: ecommerce/static/js/models/coupon_model.js [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+                    client_username: this.get('client_username'),
+                    start_date: moment.utc(this.get('start_date')),
+                    end_date: moment.utc(this.get('end_date')),
+                    stock_record_ids: stockRecordIds,
+                    code: this.get('code') || '',
+                    voucher_type: this.get('voucher_type'),
+                    quantity: parseInt(this.get('quantity'))
+                };
+
+                if (data.voucher_type !== 'Single use' || _.isEmpty(this.get('quantity'))) {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: This seems like it belongs in validation. You generally shouldn't be setting the value of fields exposed to the user in the save method. Set the default to one, and add validation to ensure it is always a positive integer.

File: ecommerce/static/js/models/coupon_model.js [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+                    data.quantity = 1;
+                }
+
+                // Enrolment code always gives 100% discount
+                switch (this.get('code_type')) {
+                case 'enrollment':
+                    data.price = parseFloat(this.get('price'));
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Why is price necessary if the discount is 100%?

File: ecommerce/static/js/models/coupon_model.js [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+                    data.benefit_type = 'Percentage';
+                    data.benefit_value = 100;
+                break;
+                case 'discount':
+                    data.price = 0;
+                    data.benefit_type = this.get('benefit_type');
+                    data.benefit_value = parseFloat(this.get('benefit_value'));
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Do not use the float data type to represent currency. Leave it as a string.

File: ecommerce/static/js/routers/course_router.js [hunk 30]
```diff
@@ -25,14 +21,7 @@ define([
 
             routes: {
                 '(/)': 'index',
-                'new(/)': 'new',
-                '*path': 'notFound'
```
Comment (clintonb) [inline ecommerce/static/js/routers/course_router.js:29]: Unless you extend the base class' `routes`, you've broken the not found route.

File: ecommerce/static/js/routers/page_router.js (hunk 20 above)
Comment (clintonb) [inline ecommerce/static/js/routers/page_router.js:64]: That TODO was put there by me when the CAT was created. It should not hold up this PR.

File: ecommerce/static/js/test/specs/views/coupon_create_view_spec.js [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+            });
+
+            it('should throw an error if submitted with blank fields', function () {
+                var errorHTML = '<strong>Error!</strong> Please complete all required fields.';
+                view.formView.submit($.Event('click'));
+                expect(view.$el.find('.alert').length).toBe(1);
+                expect(view.$el.find('.alert').html()).toBe(errorHTML);
+            });
+
+            it('should toggle fields', function () {
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/views/coupon_create_view_spec.js]: This test case is large and somewhat confusing. Break into separate tests for each voucher type.

**Comment by clintonb** [comment]
Rebase on master to fix the migration issue.

File: ecommerce/static/js/models/coupon_model.js (hunk 28 above)
Comment (dsego) [inline ecommerce/static/js/models/coupon_model.js]: I believe this is the price you pay to purchase the enrollment codes, not the course price.

Comment (mjfrey) [inline ecommerce/static/js/models/coupon_model.js]: Yes correct, this is the price paid for the code(s).

Comment (bderusha) [inline ecommerce/static/js/models/coupon_model.js]: Can we better name this so that is clear? At the minimum let's put a comment noting that. It's not really clear that's what's happening now

Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Where is it set? Regardless, currency values should be strings, not floats.

Comment (mjfrey) [inline ecommerce/static/js/models/coupon_model.js]: What do you mean where is it set?  The price is entered by the user of the form to record the price that was paid for the coupons being created.  This price is then used on the backend order creation and ultimately recorded in the invoice.

File: ecommerce/static/js/routers/page_router.js (hunk 20 above)
Comment (clintonb) [inline ecommerce/static/js/routers/page_router.js:64]: ECOM-3185

File: ecommerce/static/js/models/coupon_model.js (hunk 28 above)
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: I see no price field in the UI: https://cloud.githubusercontent.com/assets/578557/11776854/9a76b9ca-a24a-11e5-835b-03c0cf685c57.png

Comment (dsego) [inline ecommerce/static/js/models/coupon_model.js]: My apologies, the screenshot is for discounts. The price field shows up when enrollment code is selected.

File: ecommerce/static/js/routers/course_router.js (hunk 30 above)
Comment (AlasdairSwan) [inline ecommerce/static/js/routers/course_router.js:29]: @dsego I agree with @clintonb, it would be great if this could be added to the PageRouter and inherited from there.

**Comment by dsego** [comment]
FYI, latest updates. @clintonb @AlasdairSwan @bderusha

File: ecommerce/static/templates/coupon_list.html [hunk 32]
```diff
@@ -0,0 +1,12 @@
...
+    <h1 class="hd-1 emphasized">
+        <%- gettext('Coupon Codes') %>
+        <div class="pull-right">
+            <a href="/coupons/new/" class="btn btn-primary btn-small"><%- gettext('Add New Coupon') %></a>
+        </div>
+    </h1>
+</div>
+
+<table id="couponTable" class="copy copy-base table table-striped table-bordered" cellspacing="0">
+    <caption class="sr-only">Current Coupon Codes</caption>
```
Comment (bderusha) [inline ecommerce/static/templates/coupon_list.html]: needs `gettext`

**Comment by bderusha** [comment]
One small comment.  Otherwise looks good to me

File: ecommerce/static/js/views/form_view.js [hunk 33]
```diff
@@ -0,0 +1,177 @@
...
+
+                this.model.save({
+                    complete: function () {
+                        // Restore the button text
+                        $submitButton.text(btnDefaultText);
+
+                        // Re-enable the buttons
+                        $buttons.removeAttr('disabled').removeClass('disabled');
+                    },
+                    success: this.saveSuccess.bind(this),
```
Comment (clintonb) [inline ecommerce/static/js/views/form_view.js:142]: This path is not tested.

File: ecommerce/static/js/views/coupon_form_view.js [hunk 34]
```diff
@@ -0,0 +1,233 @@
...
+                }
+                else {
+                    formGroup('[name=quantity]').addClass('hidden');
+                }
+            },
+
+            /**
+             * Fill seat type options from course ID.
+             */
+            fillFromCourse: _.debounce(function () {
```
Comment (clintonb) [inline ecommerce/static/js/views/coupon_form_view.js:188]: This needs to be tested.

**Comment by dsego** [comment]
@clintonb I updated the tests.

File: ecommerce/coupons/tests/test_views.py [hunk 35]
```diff
@@ -0,0 +1,33 @@
+import json
+
+from django.conf import settings
+from django.core.urlresolvers import reverse
+import httpretty
+from testfixtures import LogCapture
+
+from ecommerce.tests.testcases import TestCase
+
+LOGGER_NAME = 'ecommerce.coupons.views'
```
Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py]: This is not used.

File: ecommerce/static/js/collections/coupon_collection.js [hunk 36]
```diff
@@ -0,0 +1,20 @@
...
+    function (Backbone,
+              BackboneSuper,
+              _,
+              PaginatedCollection,
+              CouponModel) {
+        'use strict';
+
+        return PaginatedCollection.extend({
+            model: CouponModel,
+            url: '/api/v2/products/?product_class__name=Coupon'
```
Comment (clintonb) [inline ecommerce/static/js/collections/coupon_collection.js]: Does this actually work?

You might want to consider overriding `fetch()`: http://stackoverflow.com/questions/9884179/backbone-js-urlroot-with-http-query-string.

File: ecommerce/static/js/models/coupon_model.js [hunk 37]
```diff
@@ -0,0 +1,157 @@
...
+            isDiscountCode: function () {
+                return this.get('code_type') === 'discount' ;
+            },
+
+            initialize: function () {
+                this.on('change:voucher_type', this.changeVoucherType, this);
+            },
+
+            /**
+             * When user selects the 'Sinle use' limitation option set quanity to '1'.
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Typo in "Single".

File: ecommerce/static/js/models/coupon_model.js [hunk 38]
```diff
@@ -0,0 +1,157 @@
...
+             */
+            changeVoucherType: function (model, value) {
+                if (value === 'Single use') {
+                    this.set('quantity', 1);
+                }
+            },
+
+            /**
+             * Save the Coupon.
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Either provide more detail or remove this comment. As-is, it provides no value.

File: ecommerce/static/js/models/coupon_model.js [hunk 39]
```diff
@@ -0,0 +1,157 @@
...
+                var data;
+
+                _.defaults(options || (options = {}), {
+                    type: 'POST',
+                    // The API requires a CSRF token for all POST requests using session authentication.
+                    headers: {'X-CSRFToken': $.cookie('ecommerce_csrftoken')},
+                    contentType: 'application/json'
+                });
+
+                data = {
```
Comment (clintonb) [inline ecommerce/static/js/models/coupon_model.js]: Why not use `this.toJSON()`?

File: ecommerce/static/js/models/coupon_model.js [hunk 40]
```diff
@@ -0,0 +1,148 @@
...
+            isDiscountCode: function () {
+                return this.get('code_type') === 'discount' ;
+            },
+
+            initialize: function () {
+                this.on('change:voucher_type', this.changeVoucherType, this);
+            },
+
+            /**
+             * When user selects the 'Single use' limitation option set quanity to '1'.
```
Comment (bderusha) [inline ecommerce/static/js/models/coupon_model.js]: typo. Should be `quantity`

**Comment by dsego** [comment]
@clintonb @bderusha FYI, rebased and addressed comments.

**Comment by dsego** [comment]
@clintonb Can you take another look?

**Comment by mjfrey** [comment]
@AlasdairSwan can you take one last look.

File: ecommerce/static/js/models/coupon_model.js [hunk 41]
```diff
@@ -0,0 +1,148 @@
...
+
+                _.defaults(options || (options = {}), {
+                    type: 'POST',
+                    // The API requires a CSRF token for all POST requests using session authentication.
+                    headers: {'X-CSRFToken': $.cookie('ecommerce_csrftoken')},
+                    contentType: 'application/json'
+                });
+
+                data = this.toJSON();
+                data.start_date = moment.utc (this.get('start_date'));
```
Comment (AlasdairSwan) [inline ecommerce/static/js/models/coupon_model.js]: There should not be a space between `moment.utc` and `(this.get`

File: ecommerce/static/js/models/coupon_model.js (hunk 13 above)
Comment (AlasdairSwan) [inline ecommerce/static/js/models/coupon_model.js:124]: Was the docstring added per @clintonb's comment?

File: ecommerce/static/js/test/specs/models/coupon_model_spec.js [hunk 42]
```diff
@@ -0,0 +1,130 @@
...
+            });
+
+            describe('save', function () {
+                it('should POST enrollment data', function () {
+                    spyOn($, 'ajax');
+                    var model = new Coupon(enrollment);
+                    model.save();
+                    expect($.ajax).toHaveBeenCalled();
+                    var args = $.ajax.calls.argsFor(0);
+                    var ajaxData = JSON.parse(args[0].data);
```
Comment (AlasdairSwan) [inline ecommerce/static/js/test/specs/models/coupon_model_spec.js]: All variables should be declared at the top of the function.

File: ecommerce/static/js/test/specs/models/coupon_model_spec.js [hunk 43]
```diff
@@ -0,0 +1,130 @@
...
+                    expect(ajaxData.quantity).toEqual(1);
+                });
+
+                it('should POST discount data', function () {
+                    spyOn($, 'ajax');
+                    var model = new Coupon(discount);
+                    model.save();
+                    expect($.ajax).toHaveBeenCalled();
+                    var args = $.ajax.calls.argsFor(0);
+                    var ajaxData = JSON.parse(args[0].data);
```
Comment (AlasdairSwan) [inline ecommerce/static/js/test/specs/models/coupon_model_spec.js]: All variables should be declared at the top of the function.

File: ecommerce/static/js/test/specs/views/coupon_create_view_spec.js [hunk 44]
```diff
@@ -0,0 +1,31 @@
...
+                view = new CouponCreateEditView({ model: model, editing: false }).render();
+            });
+
+            it('should throw an error if submitted with blank fields', function () {
+                var errorHTML = '<strong>Error!</strong> Please complete all required fields.';
+                view.formView.submit($.Event('click'));
+                expect(view.$el.find('.alert').length).toBe(1);
+                expect(view.$el.find('.alert').html()).toBe(errorHTML);
+            });
+
```
Comment (AlasdairSwan) [inline ecommerce/static/js/test/specs/views/coupon_create_view_spec.js:57]: I think there should also be a test to validate a positive path (i.e. that the form submits with valid fields).

File: ecommerce/static/js/test/specs/views/coupon_form_view_spec.js [hunk 45]
```diff
@@ -0,0 +1,166 @@
...
+                                    product: 2,
+                                    partner: 1
+                                }
+                            ]
+                        }
+                    ]
+                };
+
+            function visible(selector) {
+                return !view.$el.find(selector).closest('.form-group').hasClass('hidden');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/test/specs/views/coupon_form_view_spec.js]: This test makes the assumption that the selector element is always present on the page, and if it is not on the page it will return true by default. Please update to include a check for the element's presence in the DOM (returning false if it is not, else the current check).

File: ecommerce/static/js/views/coupon_form_view.js [hunk 46]
```diff
@@ -0,0 +1,236 @@
...
+                'input[name=benefit_type]': {
+                    observe: 'benefit_type'
+                },
+                '.benefit-addon': {
+                    observe: 'benefit_type',
+                    onGet: function (val) {
+                        if (val === 'Percentage') {
+                            return '%';
+                        }
+                        else if (val === 'Fixed') {
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/coupon_form_view.js]: Nitpick: `else if` should be on same line as the if closing `}` (i.e. `} else if (`

File: ecommerce/static/js/views/coupon_form_view.js [hunk 47]
```diff
@@ -0,0 +1,236 @@
...
+                }
+
+                // The only time we allow for a generation of multiple codes is
+                // when they are of type single use.
+                if (voucherType === 'Single use') {
+                    formGroup('[name=quantity]').removeClass('hidden');
+                }
+                else {
+                    formGroup('[name=quantity]').addClass('hidden');
+                }
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/coupon_form_view.js:182]: Nitpick: in JavaScript the `else if` should be on same line as the if closing `}` (i.e. `} else if (`

File: ecommerce/static/js/views/coupon_form_view.js [hunk 48]
```diff
@@ -0,0 +1,236 @@
...
+                this._super();
+            },
+
+            toggleFields: function() {
+                var formGroup = function (sel) {
+                    return this.$el.find(sel).closest('.form-group');
+                }.bind(this);
+
+                var codeType = this.model.get('code_type');
+                var voucherType = this.model.get('voucher_type');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/coupon_form_view.js]: All variables should be declared at the top of the function with a single `var` and comma-separated variables.

File: ecommerce/static/js/views/coupon_form_view.js (hunk 34 above)
Comment (AlasdairSwan) [inline ecommerce/static/js/views/coupon_form_view.js:188]: Were tests added for this?

File: ecommerce/static/js/views/coupon_form_view.js [hunk 49]
```diff
@@ -0,0 +1,236 @@
...
+            },
+
+            /**
+             * Fill seat type options from course ID.
+             */
+            fillFromCourse: _.debounce(function () {
+                var courseId = this.$el.find('[name=course_id]').val(),
+                    course = Course.findOrCreate({id: courseId}),
+                    parseId = _.compose(parseInt, _.property('id')),
+                    self = this;
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/coupon_form_view.js]: Should not need to alias this, `_.bind()` should be used instead

File: ecommerce/static/js/views/coupon_form_view.js (hunk 34 above)
Comment (dsego) [inline ecommerce/static/js/views/coupon_form_view.js:188]: Yes, https://github.com/edx/ecommerce/pull/460/files#diff-1b9bdd172014532cf4af6b0d516acdadR74

Comment (AlasdairSwan) [inline ecommerce/static/js/views/coupon_form_view.js:188]: Great, thanks

File: ecommerce/static/templates/coupon_form.html [hunk 50]
```diff
@@ -0,0 +1,107 @@
+<div class="alerts" tabindex="-1" aria-live="polite"></div>
+
+<div class="row">
+    <div class="fields col-md-6">
+        <div class="form-group">
+            <label for="title"><%= gettext('Name') %></label>
+            <input type="text" class="form-control" name="title">
+            <p class="help-block"></p>
```
Comment (AlasdairSwan) [inline ecommerce/static/templates/coupon_form.html:8]: What are these help blocks used for? I only see deleted JS referencing them.

Comment (dsego) [inline ecommerce/static/templates/coupon_form.html:8]: I believe they display error messages.

**Comment by AlasdairSwan** [comment]
Once my comments have been addressed and the branch rebased, squashed and all tests passing it should be good to go :+1:

File: ecommerce/static/templates/coupon_form.html (hunk 50 above)
Comment (AlasdairSwan) [inline ecommerce/static/templates/coupon_form.html:8]: Cool, I was just making sure that they weren't redundant with the JS updates.

File: ecommerce/static/js/models/coupon_model.js (hunk 13 above)
Comment (dsego) [inline ecommerce/static/js/models/coupon_model.js:124]: Added and then removed per his second comment. 

> Either provide more detail or remove this comment. As-is, it provides no value.

Advice?

Comment (AlasdairSwan) [inline ecommerce/static/js/models/coupon_model.js:124]: Sorry, I don't see his second comment. I'm not sure what @clintonb was referring to here re. adding a docstring. I think it is probably fine to merge without the docstring and it could always be added in a later PR if needs be.

Comment (AlasdairSwan) [inline ecommerce/static/js/models/coupon_model.js:124]: I see what you mean now. You added a comment, then Clinton asked you to remove it :) Then it is definitely good to leave out. Sorry for the confusion.

Comment (dsego) [inline ecommerce/static/js/models/coupon_model.js:124]: Cool, thanks :)
